lib/ui5Framework/AbstractResolver.js

```diff
@@ -0,0 +1,164 @@
+const path = require("path");
+const log = require("@ui5/logger").getLogger("ui5Framework:AbstractResolver");
+
+/**
+ * Abstract Resolver
+ *
+ * @public
+ * @abstract
+ * @memberof module:@ui5/project.ui5Framework
+ */
+class AbstractResolver {
+	/**
+	 * @param {*} options options
+	 * @param {string} options.version version
+	 * @param {string} [options.cwd=process.cwd()] cwd
+	 * @param {string} [options.ui5HomeDir=os.homedir()] UI5 home dir
+	 */
+	constructor({cwd, version, ui5HomeDir}) {
+		if (!version) {
+			throw new Error(`AbstractResolver: Missing parameter "version"`);
+		}
+
+		this._ui5HomeDir = ui5HomeDir || path.join(require("os").homedir(), ".ui5");
+
+		this._cwd = cwd || process.cwd();
+		this._version = version;
+
+		// Initialize libraries information
+		this.libraries = {};
+
+		this._installedCounter = 0;
+		this._cachedCounter = 0;
+	}
+
+	async _processLibrary(libraryName, libraryMetadata, errors) {
+		// Check if library is already processed
+		if (libraryMetadata[libraryName]) {
+			return;
+		}
+		// Mark library as handled
+		libraryMetadata[libraryName] = {};
+
+		log.verbose("Processing " + libraryName);
+
+		let {metadata, install} = await this.handleLibrary(libraryName);
```

Comment:
Maybe have both return parameters as promises? So that the consumer can take care of immediately defining catch handlers?